### ISSUE TITLE
Don't use already_seen hashtable when not following links

### ISSUE THREAD

**Issue by madroach** [janestreet/async_find#1]
Hi,

as far as I can see the already_seen hashtable is only needed to prevent symlink loops. It adds to the memory consumption of async_find proportional to the number of directories. This can be a quite big amount of ram for large filesystems.

**Comment by ghost** [comment]
Do you have an alternative solution in mind?

**Comment by madroach** [comment]
Something like this:

``` diff
diff --git src/async_find.ml src/async_find.ml
index 155fb79..d047ecb 100644
--- src/async_find.ml
+++ src/async_find.ml
@@ -111,7 +111,7 @@ let close t =
     begin
       t.closed <- true;
       closedir t >>| fun () ->
-      Hashtbl.clear t.already_seen;
+      if t.options.O.follow_links then Hashtbl.clear t.already_seen;
       t.to_visit <- [];
     end
   else Deferred.unit
@@ -119,7 +119,7 @@ let close t =

 (* return None if [fn] is a directory and has already been seen, otherwise Some info *)
 let is_new t (_output_fn, _path, stats as info) =
-  if stats.Stats.kind <> `Directory then
+  if not t.options.O.follow_links || stats.Stats.kind <> `Directory then
     return (Some info)
   else begin
     let uid = (stats.Stats.dev, stats.Stats.ino) in
@@ -265,7 +265,10 @@ let create ?(options=Options.default) dir =
   {
     base = dir;
     options = options;
-    already_seen = Hashtbl.Poly.create () ~size:11;
+    already_seen =
+      if options.O.follow_links
+      then Hashtbl.Poly.create () ~size:11
+      else Obj.magic ();
     to_visit = [];
     current_dir = [];
     current_handle = `Just_created;
```

**Comment by ghost** [comment]
Looks good, can you submit a PR without the `Obj.magic`? You can make `t.already_seen` an option

**Comment by madroach** [comment]
fixed by f1e52e682741558d1d104b55714b3d38f12625a1
